### PR TITLE
Fixed : Empty Rows shown at last page in Paginated data table

### LINKED ISSUE #80421: paginatedDataTable showing empty rows at the end if rowCount is less than rowsPerPage
## Steps to Reproduce
Go to the following DartPad:
[https://dartpad.dev/embed-flutter.html?gh_owner=material-components&gh_repo=material-components-flutter&gh_path=docs/components/dartpad/data_tables/normal&gh_ref=develop](https://dartpad.dev/embed-flutter.html?gh_owner=material-components&gh_repo=material-components-flutter&gh_path=docs/components/dartpad/data_tables/normal&gh_ref=develop)

Remove 2 rows (line # 63 & 64):
```
_Row('Cell A3', 'CellB3', 'CellC3', 3),
_Row('Cell A4', 'CellB4', 'CellC4', 4),
```

**Expected results:** 
There should be no empty rows at the end!
**Actual results:** 
Screen Shot 2021-04-14 at 3 10 27 PM

### PR DESCRIPTION
**Pull request by aakash-pamnani**
Fixed empty rows at the last page in paginated data table

BEFORE:
Before

AFTER:
After

Issues resolves : #80421 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by HansMuller — COMMENTED** [review]
This will need a test. 

@Hixie - once a test is added, can you review this one?

**Comment by MarkOSullivan94** [comment]
Also noticed this when making use of `PaginatedDataTable`. Would be good to see this fixed and merged into main.

If @aakash-pamnani is busy can someone else cherry-pick their commit and then a follow up commit to ensure it's tested so this can be merged?

**Comment by aakash-pamnani** [comment]
Sorry for the delay, was busy with some work, I have added the test for SizedBox height at last page.

**Review by Piinks — COMMENTED** [review]
Hi @aakash-pamnani thanks for contributing!
This looks like a breaking change, can you alter it so it is not breaking?
Some users may prefer the default empty rows, and looking at the code it appears to be very intentional in its design. I think it's a good idea to give folks the option though. Can you update this to maintain the  current default, and add a parameter folks can set to opt-in to this new behavior?

**Comment by Piinks** [comment]
It also looks like there are a lot of failing tests here, can you take a look? Thank you!

**Comment by aakash-pamnani** [comment]
Hi @Piinks thanks for reviewing the PR, I will make the required changes soon.

And the test are failing after the last merge will also look into that.

File: packages/flutter/lib/src/material/paginated_data_table.dart [hunk 1]
```diff
@@ -112,6 +112,7 @@ class PaginatedDataTable extends StatefulWidget {
     this.controller,
     this.primary,
     this.headingRowColor,
+    this.hideEmptyLastPageRows = false,
```
Comment (Piinks) [inline packages/flutter/lib/src/material/paginated_data_table.dart:115]: Hmm.. naming is hard. Having a negating name default to false can be confusing for users.

What about ..
- showEmptyRows 
- displayEmptyRows
- populateEmptyRows
- includeBlanksRows

I think having a default value of true along with a name to match would be nice. :)

File: packages/flutter/lib/src/material/paginated_data_table.dart [hunk 2]
```diff
@@ -289,6 +290,9 @@ class PaginatedDataTable extends StatefulWidget {
    /// {@macro flutter.material.dataTable.headingRowColor}
   final MaterialStateProperty<Color?>? headingRowColor;
 
+  /// If true, the last page will not show empty rows.
```
[the remark this answers is not in the document]
Comment (aakash-pamnani) [reply packages/flutter/lib/src/material/paginated_data_table.dart:293]:
```dart
/// A boolean variable that controls the visibility of empty rows on the last page
/// of a paginated data table.
///
/// The default value is `true`, which means empty rows will be visible on the
/// last page of the paginated data table. When set to `false`, empty rows will
/// not be visible on the last page.
///
bool? showEmptyRows;
```

Will this work?

**Review by Piinks — APPROVED** [review]
Flutter_LGTM
Thank you for the contribution!

**Review by cbracken — APPROVED** [review]
lgtm!

**Comment by aakash-pamnani** [comment]
> Flutter_LGTM Thank you for the contribution!

Welcome,
And ThankYou for your help and support.

**Comment by Katsumoto-crypto** [comment]
Hello, I know this issue is closed, but I'm still encountering the problem regarding the empty rows at the end of the page.

      if (widget.showEmptyRows) {
        row ??= null;
      }
      
I found out that by setting the row in here to null then the last rows are not shown anymore.
Should this issue be reopened, or should I create a new one?

**Pull request merged**